### PR TITLE
Some lemmas about dependent equality

### PR DESCRIPTION
**Pull request by herbelin**
These are various basic lemmas about dependent equality which I needed at some time to work in HoTT.

Of course, this raises a few design issues.
- Some of the `init` library uses letters `a`, `a'` for objects in `A`, other parts use `x` and `y`. Looks like a uniformity is here vain. Any opinion about that?
- I'm dreaming of a standard notation for `existT _ x y`. In an ideal world (which I hope we are step by step preparing) `(x,y)` could just work, but currently, it is not. I know Jason has some notation for `existT`. I decided here to use `{{x,y}}` but, without exporting it, so that it is not too invasive, considering that it remains an unconventional notation.
- Implicit arguments: some existing theorems are missing useful implicit arguments so that they can be used as terms. Shall we give implicit arguments to the new lemmas, even if inconsistent with the existing usage, or not. (One more thing from which a revised library could benefit: better implicit arguments.)

### CODE REVIEW

File: theories/Init/Logic.v
Comment (JasonGross) [inline theories/Init/Logic.v]: Why `trivial` rather than `constructor` or `reflexivity`, both of which seem more guaranteed to give a particular proof term?

File: theories/Init/Logic.v
Comment (JasonGross) [inline theories/Init/Logic.v]: I'd have used `f_equal2_dep`, but I don't want to delay this PR with bikeshedding, so I won't object to `f_equal_dep2`.

File: theories/Init/Specif.v
Comment (JasonGross) [inline theories/Init/Specif.v]: In HoTT, we use `( x ; y )`, which is fewer characters than `{{ x , y }}`.  Also, does this notation interfere with `{` for implicits?

File: theories/Init/Specif.v
Comment (JasonGross) [inline theories/Init/Specif.v]: Why `mk` rather than `Build` (as is autogenerated for records)?  Also, this should be `sigT`, not `sig`.  And this already exists in #385, with a different name.

**Review by JasonGross — DISMISSED** [review]
This duplicates some of the definitions in #384 and #385, with different names, possibly conflicting with them.  I'd like to get #385 merged in 8.7; would it be possible to get that merged and then put this PR on top of that?

**Comment by herbelin** [comment]
@jasongross: I'm very sorry, I forgot about #384 and #385. When on last Friday, while doing some experimentations with dependent injection, I came across some lemmas I had needed for my formalization of semi-simplicial types, I just naively looked at the `Init` files what was there thinking it was up-to-date and I added what was not there. Your proposal has precedence and I'll adjust my list after your commit is merged. My point was not to favor my version but to have the lemmas available.

@maximedenes: Do you have a schedule in mind to merge Jason's PRs? (Feel free to ask for help if this allows you to concentrate on the tasks which are most prioritary to you.)

@jasongross: about your comments: `trivial`, yes, I missed this one. `f_equal2_dep`, I guess `f_equal2` would be an equivalent of `f_equal_dep` for `f_equal2`, but here it is not the case. The name `f_equal_dep2` is not necessarily good though. About `mk_sig_eq` and `eq_sig`, I now remember the discussion. I'll try to use `eq_sig` in my formalization of semi-simplicial sets to see if it works the same.

**Comment by Zimmi48** [comment]
I think this PR was in the category: @ppedrot is not in favor of adding this to the standard library but is OK to merge it this time since other similar PRs by @JasonGross were merged.
If that is still correct, then this is just awaiting a rebase (also @JasonGross, you should remove your request for changes if your comments were taken into account).

**Comment by Zimmi48** [comment]
@herbelin Do you still want this to be merged? If yes, you should rebase it. If not, let's close.

**Comment by herbelin** [comment]
> Do you still want this to be merged?

I still think that there are a couple of important mini-improvements to do to the `Specif` library. Some of them actually include notation choices (e.g. moving to a `sigma x y z, P` notation for `{ x & { y & { z & P } } }`, or providing `p.1` notations by default, ...).

Maybe, in the absence of feedback, should I simply proceed.

In a first step, I'm going to rebase.

**Comment by herbelin** [comment]
Well, rebase (and thus synchronization with #385) done. That shall allow to think at it.

**Comment by herbelin** [comment]
This PR is a good case study about finding a "sustainable" naming policy, if possible. Let's collect all our experiences about library design to avoid making mistakes on which it is difficult to come back later (unless using a translator which is also an alternative to get rid of mistake of the past).

The question is also about notations. E.g. I refrained myself to use short notations for `eq_sym`, `eq_trans`, `f_equal` but standard notations like `⁻¹`, `⚬`, ``f`p``, could probably improve the reading (even if requiring an extra step to first be learnt).

**Comment by ejgallego** [comment]
Needs to be rebased due to multiple changes in CI.

**Review by JasonGross — APPROVED** [review]

**Comment by Zimmi48** [comment]
Would require yet another rebase for CI.

**Comment by herbelin** [comment]
Rebased to make CompCert green, even though the error is a path configuration problem.

Even if many questions remain open relatively to the naming scheme, notations, organization of the Init library, the lemmas added by this PR are useful and without apparent danger. @ppedrot, as a CODEOWNER of `Init`, I'm assigning it to you. Don't hesitate to unassign if not willing!

**Review by ppedrot — COMMENTED** [review]
As mentioned above I have mixed feelings about extending the stdlib, in particular Init. But well it seems that there is a weak consensus, so I'll merge it soon if nobody complains...

**Comment by herbelin** [comment]
> As mentioned above I have mixed feelings about extending the stdlib, in particular Init.

What do you suggest then?

**Comment by ppedrot** [comment]
My take on the stdlib is that we should decide some moratorium on its expansion and wait for a full battleplan for stdlib2. Otherwise we'll end up with half-baked solutions, inconsistent styles and non-generic enough statements.

**Comment by Zimmi48** [comment]
>My take on the stdlib is that we should decide some moratorium on its expansion and wait for a full battleplan for stdlib2.

You also said there should be a moratorium on Ltac while waiting for Ltac2, but Ltac2 is taking a long time to mature and provide a replacement so small updates on Ltac have been done meanwhile. It doesn't seem wrong to proceed in the same way on the library.

**Comment by ppedrot** [comment]
> It doesn't seem wrong to proceed in the same way on the library.

Still, we shouldn't perform this in our basement. This kind of changes should be publicly debated with major Coq users in the large, not just a handful of core devs.

**Comment by herbelin** [comment]
> This kind of changes should be publicly debated with major Coq users in the large, not just a handful of core devs.

Certainly, but how to proceed concretely? Asking for comments on coq-club? Organizing a workshop? Setting a library user/developer team? Other?

**Comment by ejgallego** [comment]
> Certainly, but how to proceed concretely? Asking for comments on coq-club? Organizing a workshop? Setting a library user/developer team? Other?

@herbelin , another possibility is to use the upcoming coq package system to allow this kind of evolution without a lot of disruption. IMO, things have advanced enough that we could propose an internship or a GSOC project.

**Comment by ppedrot** [comment]
Let's merge, we'll have more in-depth discussions later on.

**Pull request merged**
